src/value-parser.ts

```diff
@@ -14,702 +15,764 @@ const PLP_NULL = Buffer.from([0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF]);
-    } else {
-      const low = type.readUInt32LE(0);
-      const high = type.readUInt32LE(4);
-
-      if (high >= (2 << (53 - 32))) {
-        console.warn('Read UInt64LE > 53 bits : high=' + high + ', low=' + low);
-      }
+function readNChars(buf: Buffer, offset: number, dataLength: number): Result<string> {
+  if (buf.length < offset + dataLength) {
+    throw new NotEnoughDataError(offset + dataLength);
+  }
 
-      const expectedLength = low + (0x100000000 * high);
-      return readMaxKnownLength(parser, expectedLength, callback);
-    }
-  });
+  return new Result(buf.toString('ucs2', offset, offset + dataLength), offset + dataLength);
 }
 
-function readMaxKnownLength(parser: Parser, totalLength: number, callback: (value: null | Buffer) => void) {
-  const data = Buffer.alloc(totalLength, 0);
+async function readPLPStream(parser: Parser): Promise<null | Buffer[]> {
+  while (parser.buffer.length < parser.position + 8) {
+    await parser.waitForChunk();
+  }
+
+  const type = parser.buffer.slice(parser.position, parser.position + 8);
+  parser.position += 8;
+
+  if (type.equals(PLP_NULL)) {
+    return null;
+  }
 
+  const chunks: Buffer[] = [];
   let offset = 0;
-  function next(done: any) {
-    parser.readUInt32LE((chunkLength) => {
-      if (!chunkLength) {
-        return done();
-      }
 
-      parser.readBuffer(chunkLength, (chunk) => {
-        chunk.copy(data, offset);
-        offset += chunkLength;
+  while (true) {
+    while (parser.buffer.length < parser.position + 4) {
+      await parser.waitForChunk();
+    }
 
-        next(done);
-      });
-    });
-  }
+    const chunkLength = parser.buffer.readUInt32LE(parser.position);
+    parser.position += 4;
 
-  next(() => {
-    if (offset !== totalLength) {
-      throw new Error('Partially Length-prefixed Bytes unmatched lengths : expected ' + totalLength + ', but got ' + offset + ' bytes');
+    if (!chunkLength) {
+      break;
     }
 
-    callback(data);
-  });
-}
+    while (parser.buffer.length < parser.position + chunkLength) {
+      await parser.waitForChunk();
+    }
 
-function readMaxUnknownLength(parser: Parser, callback: (value: null | Buffer) => void) {
-  const chunks: Buffer[] = [];
+    chunks.push(parser.buffer.slice(parser.position, parser.position + chunkLength));
+    parser.position += chunkLength;
+    offset += chunkLength;
+  }
 
-  let length = 0;
-  function next(done: any) {
-    parser.readUInt32LE((chunkLength) => {
-      if (!chunkLength) {
-        return done();
-      }
+  if (!type.equals(UNKNOWN_PLP_LEN)) {
+    const low = type.readUInt32LE(0);
+    const high = type.readUInt32LE(4);
 
-      parser.readBuffer(chunkLength, (chunk) => {
-        chunks.push(chunk);
-        length += chunkLength;
+    if (high >= (2 << (53 - 32))) {
+      console.warn('Read UInt64LE > 53 bits : high=' + high + ', low=' + low);
+    }
 
-        next(done);
-      });
-    });
+    const expectedLength = low + (0x100000000 * high);
```

Comment:
I know you copied this from the old version of this code, but I'd love if we could improve this.

This code is reading the lower 4 bytes and the higher 4 bytes of an 8 byte integer, to then try and put it back together as a 8 byte integer. I think instead of doing that, we can just read this as a bigint and then convert it back to a `Number`?

Reply:
The part for checking length looks like this now. Not sure if we still want the expectedLength var since 
`if (!type.equals(UNKNOWN_PLP_LEN)) {
    const expectedLength = Number(readBigInt(type, 0).value);
    if (currentLength !== expectedLength) {
      throw new Error('Partially Length-prefixed Bytes unmatched lengths : expected ' + expectedLength + ', but got ' + currentLength + ' bytes');
    }
  }`